network/src/noise/handshake.rs

```diff
@@ -348,35 +351,52 @@ impl NoiseUpgrader {
             .map_err(|err| NoiseHandshakeError::ServerParseClient(remote_peer_short, err))?;
 
         // if mutual auth mode, verify the remote pubkey is in our set of trusted peers
-        let trust_level = if let Some(trusted_peers) = self.auth_mode.trusted_peers() {
-            match trusted_peers.read().get(&remote_peer_id) {
-                Some(remote_pubkey_set) => {
-                    if !remote_pubkey_set.contains(&remote_public_key) {
-                        return Err(NoiseHandshakeError::UnauthenticatedClientPubkey(
+        let trust_level = match &self.auth_mode {
+            HandshakeAuthMode::Mutual { trusted_peers, .. } => {
+                match trusted_peers.read().get(&remote_peer_id) {
+                    Some(remote_pubkey_set) => {
+                        if !remote_pubkey_set.contains(&remote_public_key) {
+                            return Err(NoiseHandshakeError::UnauthenticatedClientPubkey(
+                                remote_peer_short,
+                                hex::encode(remote_public_key.as_slice()),
+                            ));
+                        }
+                    }
+                    None => {
+                        return Err(NoiseHandshakeError::UnauthenticatedClient(
                             remote_peer_short,
-                            hex::encode(remote_public_key.as_slice()),
-                        ));
+                            remote_peer_id,
+                        ))
+                    }
+                };
+                TrustLevel::Trusted
+            }
+            HandshakeAuthMode::MaybeMutual(trusted_peers) => {
+                match trusted_peers.read().get(&remote_peer_id) {
+                    Some(remote_pubkey_set) => {
+                        if !remote_pubkey_set.contains(&remote_public_key) {
+                            return Err(NoiseHandshakeError::UnauthenticatedClientPubkey(
```

Comment:
shouldn't you return `TrustLevel::Untrusted` instead here?

Reply:
Great question, let's brainstorm on that briefly
1) We assume that the peer id is not derived in the mutual auth case, so we would need to add that check back in
2) If a peer id is derived but it assumes mutual auth and does not get it, they won't get an error and may assume they are privileged

I think the key question is in how we instruct folks to share their id, key pairs -- I suspect we'll ask them to use their VASP (whatever partner) account ids and share a public key

Reply:
I would have thought this would be more correct:

```rust
let is_trusted = if let Some(remote_pubkey_set) = trusted_peers.read().get(&remote_peer_id) {
  if remote_pubkey_set.contains(&remote_public_key) {
    TrustLevel::Trusted
  }
} else {
  TrustLevel::Untrusted
};

let derived_remote_peer_id =
    PeerId::from_identity_public_key(remote_public_key);
if derived_remote_peer_id != remote_peer_id {
    return Err(NoiseHandshakeError::ClientPeerIdMismatch(
        remote_peer_short,
        remote_peer_id,
        derived_remote_peer_id,
    ));
}

is_trusted
```

Reply:
I would actually argue that the current approach implemented is correct. If I have a known peer id slipping into an untrusted level -- it is weird. Also if a peer is untrusted, the get statement will fail resulting in the remote_pub_keyset being None.

trusted_peers.read().get(...)

would be None

so maybe "it's late" :P